arangod/Aql/OptimizerRules.cpp

```diff
@@ -5843,14 +5843,92 @@ void arangodb::aql::optimizeTraversalsRule(Optimizer* opt,
     }
 
     outVariable = traversal->pathOutVariable();
-    if (outVariable != nullptr && !n->isVarUsedLater(outVariable)) {
-      // traversal path outVariable not used later
-      options->setProducePaths(false);
-      if (std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
-          pruneVars.end()) {
-        traversal->setPathOutput(nullptr);
+    if (outVariable != nullptr) {
+      if (!n->isVarUsedLater(outVariable)) {
+        // traversal path outVariable not used later
+        if (std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
+            pruneVars.end()) {
+          options->setProducePaths(/*vertices*/ false, /*edges*/ false, /*weights*/ false);
+          traversal->setPathOutput(nullptr);
+        } else {
+          // we still need to build the path because PRUNE relies on it
+          options->setProducePaths(/*vertices*/ true, /*edges*/ true, /*weights*/ true);
+        }
+        modified = true;
+      } else {
+        // path is used later, but lets check which of its sub-attributes
+        // "vertices" or "edges" are in use (or the complete path)
+        std::unordered_set<std::string> attributes;
+        VarSet vars;
+        bool canOptimize = true;
+
+        if (options->usesPrune()) {
+          std::vector<Variable const*> pruneVars;
+          traversal->getPruneVariables(pruneVars);
+          if (std::find(pruneVars.begin(), pruneVars.end(), outVariable) != pruneVars.end()) {
+            // prune uses the path. we don't know if the optimization
```

Comment:
I think we are able to analyze the Prune statement as well.
And figure out if / which of the items are used by prune.